fpt-egui/src/main.rs

```diff
@@ -213,100 +228,205 @@ impl FPT {
+                        egui::Grid::new("VRAM-registers-3").striped(true).show(ui, |ui| {
+                            ui.monospace("SCX");
+                            ui.monospace(format!("{:08b}", bus.scx()));
+                            ui.end_row();
+                            ui.monospace("SCY");
+                            ui.monospace(format!("{:08b}", bus.scy()));
+                            ui.end_row();
+                        });
+                    });
+                });
+                ui.add_space(20.0);
+                ui.heading("CPU");
+                ui.separator();
+                ui.horizontal_wrapped(|ui| {
+                    macro_rules! cpu_register {
+                        ($ui:expr, $high_label:literal : $high_value:expr, $low_label:literal : $low_value:expr) => {
+                            $ui.colored_label(Color32::LIGHT_BLUE, $high_label);
+                            $ui.monospace(format!("{:08b}", $high_value));
+                            $ui.code(format!("{:04X}", bitwise::word16($high_value, $low_value)));
+                            $ui.monospace(format!("{:08b}", $low_value));
+                            $ui.colored_label(Color32::LIGHT_BLUE, $low_label);
+                        }
+                    }
+                    let gb = self.gb();
+                    let cpu = gb.cpu();
+                    egui::Grid::new("cpu_registers_a-e").num_columns(4).min_col_width(10.0).striped(true).show(ui, |ui| {
+                        cpu_register!(ui, "A": cpu.a(), "F": cpu.f()); ui.end_row();
+                        cpu_register!(ui, "B": cpu.b(), "C": cpu.c()); ui.end_row();
+                        cpu_register!(ui, "D": cpu.d(), "E": cpu.e()); ui.end_row();
+                    });
+                    ui.separator();
+                    ui.vertical(|ui| {
+                        ui.horizontal(|ui| {
+                            cpu_register!(ui, "H": cpu.h(), "L": cpu.l());
+                        });
+                        ui.horizontal(|ui| {
+                            ui.colored_label(Color32::LIGHT_BLUE, "SP");
+                            ui.monospace(format!("{:016b}", cpu.sp()));
+                            ui.code(format!("{:#04X}", cpu.sp()));
+                        });
+                        ui.horizontal(|ui| {
+                            ui.colored_label(Color32::LIGHT_BLUE, "PC");
+                            ui.monospace(format!("{:016b}", cpu.pc()));
+                            ui.code(format!("{:#04X}", cpu.pc()));
+                        });
+                    });
+                });
+                ui.add_space(20.0);
+                ui.horizontal(|ui| {
+                    ui.heading("Debugger:");
+                    if ui.button("Pause").clicked() {
+                        self.paused = true;
+                    }
+                    if ui.button("Step").clicked() {
+                        self.gb().frame();
```

Comment:
this doesn't make much sense. sure, it updates some of the gameboy's state but it doesn't draw the frame or anything else. it also wont work when with interrupts, input we'll have to feed into the gameboy, audio, ...

Reply:
yeah, I should've called it "Next frame". But I didn't give much thought about the function of the buttons yet, I only wrote them to see where I liked them best in the UI, as an excuse to play with egui layouts (speaking of which, I totally forgot to showcase the Registers collapsible in the video)